### ISSUE TITLE
Add React-Query (or similar) in react components

### ISSUE THREAD

**Issue by saranshgrover** [thewca/worldcubeassociation.org#6361]
https://react-query.tanstack.com/

We currently have self-created 2 hooks for loading and saving data, `useLoadedData` and `useSaveData`. While both work well, by migrating to react-query, we would be able to easily an exhaustive API as well as caching while still maintaining almost the same functionality.

This will be blocked until #5300 and #6360 are merged, and I'd like more feedback - specifically with members who want to work on the React end about this change.

**Comment by viroulep** [comment]
Using a well tested, used, and maintained library for our queries is definitely something we should look into!
I've never looked into them, is there any other alternatives out there, and did you have a good experience using any of them?

Do you know how caching is implemented?
I've taken a look at the [`queryCache`](https://github.com/tannerlinsley/react-query/blob/master/src/core/queryCache.ts) implementation and it looks to me that it's done in-memory (but I'm very very unsure all the logic regarding its storage is here!).
It's worth keeping in mind that, at the moment, our js "frontend" is quite unusual compared to regular react applications: small components are mounted on "regular" html pages, and we don't have components with long-lasting lives (if you change pages you basically reload everything).
The only example I have in mind that would last longer than a single page is the competition results component: if my understanding of the implementation is right caching would benefit a user switching back and forth from different events on the competition's results page, but upon refreshing the page the cache would be gone.
Writing this I notice [this experimental feature](https://react-query.tanstack.com/plugins/persistQueryClient) that allows to persist the cache to local storage, which would better suit our setup.

Anyway, just some stuff to keep in mind when we do decide we want to cache some queries.

> This will be blocked until #5300 and #6360 are merged

If you want to try the integration of the library I definitely think you could do so on the [post widget](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/webpacker/javascript/posts_widget/index.js)!
There is a simple load so I'd expect the diff to switch from our hook to the library to be fairly minimal.
And if it is small it should definitely be easy to switch other components too :)

**Comment by dunkOnIT** [comment]
@FinnIckler is this still relevant?

**Comment by gregorbg** [comment]
We've certainly added it in a lot of places but we haven't fully made the switch. This would be a very nice project for someone who wants to take a "stroll" through our React codebase. Excellent second (maybe even first, if you know the basics of React) issue for frontend-minded people.

**Comment by FinnIckler** [comment]
I agree, but it's used in 44 files. So I think it's easier just changing useLoadedData to use useQuery under the hood and then replace them?
